File: packages/typescript-core/src/calls.ts
// On-chain call arguments for callers submitting with their own Substrate client;
// `@openmatter-network/matter-sdk`'s MatterClient submits them for you. Builders
// take an `Aad` tag so a secret is stored under the AAD it was sealed with.

import { Aad, aadBytes } from "./aad.js";
import type { EncryptedSecret } from "./types.js";

/** Arguments for `secrets.storeSecret(payload, epoch, label, aad)`. */
export interface StoreSecret {
  payload: EncryptedSecret;
  epoch: number;
  label: Uint8Array;
  aad: Uint8Array;
}

/** Build store args, deriving the AAD bytes from the registry tag used at seal time. */
export function storeSecret(
  payload: EncryptedSecret,
  epoch: number,
  label: Uint8Array | string,
  aad: Aad,
): StoreSecret {
  return {
    payload,
    epoch,
    label: typeof label === "string" ? new TextEncoder().encode(label) : label,
    aad: aadBytes(aad),
  };
}

/** Arguments for `secrets.rotateSecret(secret_id, payload, epoch, aad)`. */
export interface RotateSecret {
  secretId: bigint;
  payload: EncryptedSecret;
  epoch: number;
  aad: Uint8Array;
}

/** Build rotate args from the registry AAD tag used at seal time. */
export function rotateSecret(
  secretId: bigint,
  payload: EncryptedSecret,
  epoch: number,
  aad: Aad,
): RotateSecret {
  return { secretId, payload, epoch, aad: aadBytes(aad) };
}

/** Who a secret is granted to: the chain's `GrantTarget<AccountId>` enum. */
export type GrantTarget =
  /** Another user or a resource node, named by account. */
  | { readonly User: Uint8Array }
  /** A deployment: authorizes whichever resource is currently assigned to it. */
  | { readonly Deployment: bigint };

const ACCOUNT_ID_BYTES = 32;

/** A grant target naming a 32-byte account id. */
export function grantToUser(account: Uint8Array): GrantTarget {
  if (account.length !== ACCOUNT_ID_BYTES) {
    throw new Error(`account must be ${ACCOUNT_ID_BYTES} bytes, got ${account.length}`);
  }
  return { User: account };
}

/** A grant target naming a deployment. */
export function grantToDeployment(deployment: bigint): GrantTarget {
  return { Deployment: deployment };
}

/** Arguments for `secrets.grantAccess(secret_id, target)`. */
export interface GrantAccess {
  secretId: bigint;
  target: GrantTarget;
}

export function grantAccess(secretId: bigint, target: GrantTarget): GrantAccess {
  return { secretId, target };
}

/** Arguments for `secrets.revokeAccess(secret_id, target)`. */
export interface RevokeAccess {
  secretId: bigint;
  target: GrantTarget;
}

/** Build revoke args. The target must match the grant exactly, or this is a no-op. */
export function revokeAccess(secretId: bigint, target: GrantTarget): RevokeAccess {
  return { secretId, target };
}

/** Arguments for `secrets.deleteSecret(secret_id)`. Owner only, irreversible. */
export interface DeleteSecret {
  secretId: bigint;
}

export function deleteSecret(secretId: bigint): DeleteSecret {
  return { secretId };
}

File: bindings/python/python/matter_sdk/calls.py
"""substrate-interface ``call_params`` for callers who submit with their own client;
:class:`~matter_sdk.client.MatterClient` submits them for you.

Builders take an ``Aad`` tag so a secret cannot be stored under a different AAD
than it was sealed with.
"""

from typing import Tuple, Union

from .aad import Aad, aad_bytes
from .hexutil import to_hex

# (binding_id, capsule, proof, ct), as ``encrypt`` returns.
EncryptedSecret = Tuple[bytes, bytes, bytes, bytes]


def _payload(env: EncryptedSecret) -> dict:
    binding_id, capsule, proof, ct = env
    return {
        "binding_id": to_hex(binding_id),
        "capsule": to_hex(capsule),
        "proof": to_hex(proof),
        "ct": to_hex(ct),
    }


def store_secret(
    payload: EncryptedSecret, epoch: int, label: Union[str, bytes], aad: Union[Aad, str, bytes]
) -> dict:
    """``call_params`` for ``secrets.store_secret(payload, epoch, label, aad)``."""
    return {
        "payload": _payload(payload),
        "epoch": int(epoch),
        "label": to_hex(label.encode() if isinstance(label, str) else bytes(label)),
        "aad": to_hex(aad_bytes(aad)),
    }


def rotate_secret(
    secret_id: int, payload: EncryptedSecret, epoch: int, aad: Union[Aad, str, bytes]
) -> dict:
    """``call_params`` for ``secrets.rotate_secret(secret_id, payload, epoch, aad)``."""
    return {
        "secret_id": int(secret_id),
        "payload": _payload(payload),
        "epoch": int(epoch),
        "aad": to_hex(aad_bytes(aad)),
    }


ACCOUNT_ID_BYTES = 32


def grant_to_user(account: bytes) -> dict:
    """A ``GrantTarget`` naming an account."""
    if len(account) != ACCOUNT_ID_BYTES:
        raise ValueError(f"account must be {ACCOUNT_ID_BYTES} bytes, got {len(account)}")
    return {"User": to_hex(account)}


def grant_to_deployment(deployment: int) -> dict:
    """A ``GrantTarget`` naming a deployment: authorizes whichever resource is
    currently assigned to it."""
    return {"Deployment": int(deployment)}


def grant_access(secret_id: int, target: dict) -> dict:
    """``call_params`` for ``secrets.grant_access(secret_id, target)``.

    ``target`` is a ``GrantTarget`` enum from :func:`grant_to_user` or
    :func:`grant_to_deployment`, not a bare account id.
    """
    return {"secret_id": int(secret_id), "target": target}


def revoke_access(secret_id: int, target: dict) -> dict:
    """``call_params`` for ``secrets.revoke_access(secret_id, target)``.

    The target must match the grant exactly, or this is a no-op on chain.
    """
    return {"secret_id": int(secret_id), "target": target}


def delete_secret(secret_id: int) -> dict:
    """``call_params`` for ``secrets.delete_secret(secret_id)``.

    Removes the secret and every grant on it. Owner only, and irreversible.
    """
    return {"secret_id": int(secret_id)}
